src/storage/src/hummock/shared_buffer/mod.rs

```diff
@@ -17,22 +17,67 @@ pub mod shared_buffer_batch;
 #[allow(dead_code)]
 pub mod shared_buffer_uploader;
 
-use std::collections::HashMap;
+use std::collections::{BTreeMap, HashMap};
 use std::ops::RangeBounds;
 
 use itertools::Itertools;
+use risingwave_hummock_sdk::is_remote_sst_id;
+use risingwave_hummock_sdk::key::user_key;
+use risingwave_pb::hummock::{KeyRange, SstableInfo, VNodeBitmap};
 
 use self::shared_buffer_batch::SharedBufferBatch;
-use crate::hummock::shared_buffer::shared_buffer_batch::IndexedSharedBufferBatches;
 use crate::hummock::shared_buffer::shared_buffer_uploader::{UploadTaskId, UploadTaskPayload};
-use crate::hummock::utils::range_overlap;
+use crate::hummock::utils::{filter_single_sst, range_overlap};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum UncommittedData {
```

Comment:
What is the benefit of putting `SstableInfo` and `SharedBufferBatch` in the same emum? Looks like we need to do extra checks on the enum type outside when using `UncommittedData`.

Reply:
In the future when we support write any time, we will have to maintain the order of SSTs and write batches together, so the code will be easier if we store SSTs and write batches together as an `UncommittedData`. If we do not introduce such `UncommittedData` enum, when we handle `get_overlap_data`, we will have to return SSTs and write batches separately, and when we build the ordered merge iterator, we will have to merge the two ordered array again, which makes the code more complicated and less readable. 

Another seemingly feasible solution will be maintaining the order within SSTs and the order within write batches separately and assume that the order of write batches is always later than SSTs. But assumption will limit our ability to concurrently upload write batches. Let's assume that we first have write batch `w1` and `w2`, and we issue a compaction task `t1` to compact `w1` and `w2` into SST `s1`. And then when we have some other write batch `w3` and `w4` coming and take up too much memory, we will want to issue another compaction task `t2` to compact them into `s2`. However, since we have the assumption that the order of write batches are later than SSTs, `t2` will have to wait for `t1` to finish. This is because that once `t2` is issued, it can finish earlier than `t1`, and then `s2` will exist while `w1` and `W2` still exist, and this breaks the assumption. Therefore, such assumption limits our ability to issue concurrent compaction task. And without this assumption, the order of SSTs and write batches will be mixed together, and there is no good to maintain them separately.

Reply:
I get it now. Thanks for the explanation.